external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -2512,7 +2515,30 @@ fn exp(context: &mut Context, e: Box<E::Exp>) -> Box<N::Exp> {
                             .env
                             .add_diag(diag!(TypeSafety::InvalidCallTarget, (mloc, msg)));
                     }
-                    NE::VarCall(v, nes)
+                    // If this variable refers to a local (num > 0) or it isn't syntax, error.
+                    if !v.value.is_syntax_identifier() {
+                        let msg = format!(
+                            "Unexpected invocation of parameter or local '{}'. \
+                                     Non-syntax variables cannot be invoked as functions.",
+                            v.value.name
+                        );
+                        context
+                            .env
+                            .add_diag(diag!(TypeSafety::InvalidCallTarget, (v.loc, msg)));
+                        NE::UnresolvedError
+                    } else if v.value.id != 0 {
+                        let msg = format!(
+                            "Unexpected invocation of non-parameter variable '{}'. \
+                                     Only lambda-typed syntax parameters may be invoked",
+                            v.value.name
+                        );
+                        context
+                            .env
+                            .add_diag(diag!(TypeSafety::InvalidCallTarget, (v.loc, msg)));
+                        NE::UnresolvedError
```

Comment:
When does this occur?

Reply:
You can't actually write it today, because naming won't resolve the binding in the `let` form, but if the compiler (reasonable) raised an error and kept going at `let $x = 0`, this would catch that and report an error as well.